front_end/entrypoints/inspector_main/CohtmlPanel.ts

```diff
@@ -0,0 +1,259 @@
+
+  captureRendDesc: 'Capture Rend File',
+
+  capturePageTitle: 'Capture Full Page',
+
+  capturePageDesc: 'Capture Full Page',
+
+  clearCachedUnusedImagesTitle: 'Clear Cached Unused Images',
+
+  clearCachedUnusedImagesDesc: 'Removes all unused images (raster and svg) from internal caches.',
+
+  getSystemCacheTitle: 'Get System Cache Stats',
+
+  getSystemCacheDesc: 'Get statistics for the system-wide caches',
+
+};
+const str_ = i18n.i18n.registerUIStrings('entrypoints/inspector_main/CohtmlPanel.ts', UIStrings);
+const i18nString = i18n.i18n.getLocalizedString.bind(undefined, str_);
+
+let cohtmlPanelViewInstance: CohtmlPanelView;
+
+export class CohtmlPanelView extends UI.Widget.VBox implements SDK.TargetManager.SDKModelObserver<SDK.CohtmlDebugModel.CohtmlDebugModel>{
+
+  private drawMetaDataSetting: Common.Settings.Setting<any>;
+
+  private continuousRepaintSetting: Common.Settings.Setting<any>;
+
+  private cohtmlDebugModel?: SDK.CohtmlDebugModel.CohtmlDebugModel|null;
+
+  // overlayAgent: ProtocolProxyApi.OverlayApi;
+
+  private constructor() {
+    super(true);
+    this.registerRequiredCSS('entrypoints/inspector_main/renderingOptions.css');
+
+    SDK.TargetManager.TargetManager.instance().observeModels(SDK.CohtmlDebugModel.CohtmlDebugModel, this);
+
+    let mainTarget = SDK.TargetManager.TargetManager.instance().mainTarget();
+    if (mainTarget)
+    {
+      this.cohtmlDebugModel = mainTarget.model(SDK.CohtmlDebugModel.CohtmlDebugModel);
+    }
+
+    this.drawMetaDataSetting = Common.Settings.Settings.instance().moduleSetting('drawMetaData');
+    this.continuousRepaintSetting = Common.Settings.Settings.instance().moduleSetting('continuousRepaint');
+
+    this.appendCheckbox(
+        i18nString(UIStrings.paintFlashing), i18nString(UIStrings.highlightsAreasOfThePageGreen),
+        Common.Settings.Settings.instance().moduleSetting('showPaintRects'));
+    this.appendCheckbox(
+        i18nString(UIStrings.redrawFlashing), i18nString(UIStrings.highlightsAreasOfThePageRed),
+        Common.Settings.Settings.instance().moduleSetting('showRedrawRects'));
+
+    this.contentElement.createChild('div').classList.add('panel-section-separator');
+
+    this.appendCheckbox(UIStrings.toggleMetaDataTitle, UIStrings.toggleMetaDataDesc,
+                        Common.Settings.Settings.instance().moduleSetting('drawMetaData'));
+
+    this.appendCheckbox(UIStrings.toggleContinuousRepaintTitle, UIStrings.toggleContinuousRepaintDesc,
+                        Common.Settings.Settings.instance().moduleSetting('continuousRepaint'));
+
+    this.createButton(UIStrings.dumpDomTitle, UIStrings.dumpDomDesc, () => {
+      if (this.cohtmlDebugModel)
+      {
+        this.cohtmlDebugModel.dumpDOM();
+      }
+    });
+
+    this.createButton(UIStrings.dumpStackingContextTitle, UIStrings.dumpStackingContextDesc, () => {
+      if (this.cohtmlDebugModel)
+      {
+        this.cohtmlDebugModel.dumpStackingContext();
+      }
+    });
```

Comment:
A bit of a bikeshed, but if you want we can make a map for all the buttons which might make future additions easier.

```JS
const buttons = [
  { title: UIStrings.dumpStackingContextTitle, description: UIStrings.dumpStackingContextDesc, method: "dumpStackingContext" },
  { title: UIStrings.dumpUsedImagesTitle, description: UIStrings.dumpUsedImagesDesc, method: "dumpUsedImages" }
    // ...
];
for (const button of buttons) {
  this.createButton(button.title, button.description, () => {
    if (this.cohtmlDebugModel && this.cohtmlDebugModel[button.method]) {
      this.cohtmlDebugModel[button.method]();
    }
  });
}
```
Or the following which should make delimiters easier.
```JS
const createButton = (title, description, method) => {
  this.createButton(title, description, () => {
    if (this.cohtmlDebugModel && this.cohtmlDebugModel[method]) {
      this.cohtmlDebugModel[method]();
    }
  });
};
createButton(UIStrings.dumpStackingContextTitle, UIStrings.dumpStackingContextDesc, "dumpStackingContext");
createButton(UIStrings.dumpUsedImagesTitle, UIStrings.dumpUsedImagesDesc, "dumpUsedImages");
```

Reply:
ok, how do I do this in TypeScript in a type safe way?

Reply:
I've done it through liberal use of `any`